### DISCUSSION TITLE
No Event are being Listened When subscription states are changed

### DISCUSSION THREAD

**Discussion by HilalLko** [imdhemy/laravel-in-app-purchases#13] [Q&A] [answered]
Following up with the example provided to handle subscription purchase events.
Created a new Listener for "SubscriptionRenewed" as Well as "SubscriptionCanceled" events.
But when the subscription renewed nothing happens nor got any logs.

**Code in my "AutoRenewSubscriptionPlayStore" Listener**
```php
class AutoRenewSubscriptionPlayStore
{    
    /**
     * Handle the event.
     *
     * @param  object  $event
     * @return void
     */
    public function handle(SubscriptionRenewed $event)
    {
        Log::info($event);
        $notification = $event->getServerNotification();
        $subscription = $notification->getSubscription();
        $uniqueIdentifier = $subscription->getUniqueIdentifier();
        $expirationTime = $subscription->getExpiryTime();
        $userId = User::findUserBySubscriptionId($uniqueIdentifier,'PlayStore');
        $user = User::find($userId);
        $subscription = UserSubscription::where('subscription_id',$uniqueIdentifier)->where('user_id',$userId)->first();
        $subscription->subscription_ends = $expirationTime;
        $subscription->save();
        $file = (isset($user->profile_picture)) ? $user->profile_picture : (($user->mediaImages->count() > 0) ? $user->profile_image : '');
        $image = generateS3FilePath($file);
        event(new SendPushNotification($user->id, trans('notification.subscription_updated'), array(trans('notification.subscription_updated'),'Subscription Updated',$image)));
    }
}
```
@imdhemy Help needed

**Comment by imdhemy** [comment, marked as answer]
Did you make all the following steps?
## Google Play
1. Configure RTDN. [Configure Real-time developer notifications](https://developer.android.com/google/play/billing/getting-ready#configure-rtdn)
2. Use the specified route for the RTDN, [more information here](https://github.com/imdhemy/laravel-in-app-purchases#i1-routing).
3. Add the listener to the config file `purchase.php`, [attach the created listener](https://github.com/imdhemy/laravel-in-app-purchases#configuration) to the proper event.

## App Store
1. Configure the App store [server notifications](https://developer.apple.com/documentation/appstoreservernotifications/enabling_app_store_server_notifications).
2. Use the specified route for the App store server notifications, [more information here](https://github.com/imdhemy/laravel-in-app-purchases#i1-routing).
3. Add the listener to the config file `purchase.php`, [attach the created listener](https://github.com/imdhemy/laravel-in-app-purchases#configuration) to the proper event.

## Summary
The config file `purchase.php`, contains two separate groups for different purchase events, be sure to attach the listener to the proper event for [google](https://github.com/imdhemy/laravel-in-app-purchases/blob/3cd1ab5a58b81e775df7c905e92a1696a3ca05cc/config/purchase.php#L39) or [apple](https://github.com/imdhemy/laravel-in-app-purchases/blob/3cd1ab5a58b81e775df7c905e92a1696a3ca05cc/config/purchase.php#L58) after setting up the server push notifications on the provider console.

**Reply by HilalLko** [reply]
Yeah but only for Google Play 

Getting this in logs now

[2020-12-11 06:15:58] staging.ERROR: array_values() expects parameter 1 to be array, null given {"exception":"[object] (ErrorException(code: 0): array_values() expects parameter 1 to be array, null given at /var/app/current/vendor/imdhemy/google-play-billing/src/DeveloperNotifications/DeveloperNotification.php:76)
[stacktrace]
#0 [internal function]: Illuminate\\Foundation\\Bootstrap\\HandleExceptions->handleError(2, 'array_values() ...', '/var/app/curren...', 76, Array)
#1 /var/app/current/vendor/imdhemy/google-play-billing/src/DeveloperNotifications/DeveloperNotification.php(76): array_values(NULL)
#2 /var/app/current/vendor/imdhemy/laravel-purchases/src/Http/Controllers/ServerNotificationController.php(24): Imdhemy\\GooglePlay\\DeveloperNotifications\\DeveloperNotification::parse('ZnNkZnNkZnNk')
#3 /var/app/current/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): Imdhemy\\Purchases\\Http\\Controllers\\ServerNotificationController->google(Object(Imdhemy\\Purchases\\Http\\Requests\\GoogleDeveloperNotificationRequest))
#4 /var/app/current/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(45): Illuminate\\Routing\\Controller->callAction('google', Array)
#5 /var/app/current/vendor/laravel/framework/src/Illuminate/Routing/Route.php(219): Illuminate\\Routing\\ControllerDispatcher->dispatch(Object(Illuminate\\Routing\\Route), Object(Imdhemy\\Purchases\\Http\\Controllers\\ServerNotificationController), 'google')
#6 /var/app/current/vendor/laravel/framework/src/Illuminate/Routing/Route.php(176): Illuminate\\Routing\\Route->runController()
#7 /var/app/current/vendor/laravel/framework/src/Illuminate/Routing/Router.php(681): Illuminate\\Routing\\Route->run()
#8 /var/app/current/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(130): Illuminate\\Routing\\Router->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))
#9 /var/app/current/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(105): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#10 /var/app/current/vendor/laravel/framework/src/Illuminate/Routing/Router.php(683): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#11 /var/app/current/vendor/laravel/framework/src/Illuminate/Routing/Router.php(658): Illuminate\\Routing\\Router->runRouteWithinStack(Object(Illuminate\\Routing\\Route), Object(Illuminate\\Http\\Request))
#12 /var/app/current/vendor/laravel/framework/src/Illuminate/Routing/Router.php(624): Illuminate\\Routing\\Router->runRoute(Object(Illuminate\\Http\\Request), Object(Illuminate\\Routing\\Route))
#13 /var/app/current/vendor/laravel/framework/src/Illuminate/Routing/Router.php(613): Illuminate\\Routing\\Router->dispatchToRoute(Object(Illuminate\\Http\\Request))
#14 /var/app/current/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(170): Illuminate\\Routing\\Router->dispatch(Object(Illuminate\\Http\\Request))
#15 /var/app/current/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(130): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}(Object(Illuminate\\Http\\Request))
#16 /var/app/current/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#17 /var/app/current/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#18 /var/app/current/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#19 /var/app/current/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#20 /var/app/current/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#21 /var/app/current/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#22 /var/app/current/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(63): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#23 /var/app/current/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Illuminate\\Foundation\\Http\\Middleware\\CheckForMaintenanceMode->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#24 /var/app/current/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#25 /var/app/current/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(171): Fideloper\\Proxy\\TrustProxies->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#26 /var/app/current/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(105): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#27 /var/app/current/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(145): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#28 /var/app/current/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(110): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter(Object(Illuminate\\Http\\Request))
#29 /var/app/current/public/index.php(55): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))
#30 {main}
"}

**Reply by HilalLko** [reply]
As well as 

/var/log/nginx/access.log
----------------------------------------
172.31.1.124 - - [11/Dec/2020:06:28:26 +0000] "POST /purchases/subscriptions/google HTTP/1.1" 500 11696 "-" "APIs-Google; (+https://developers.google.com/webmasters/APIs-Google.html)" "35.187.129.246"
172.31.1.124 - - [11/Dec/2020:06:28:27 +0000] "POST /purchases/subscriptions/google HTTP/1.1" 500 8546 "-" "APIs-Google; (+https://developers.google.com/webmasters/APIs-Google.html)" "35.187.129.246"

**Reply by imdhemy** [reply]
This is not a valid RTDN notification payload `'ZnNkZnNkZnNk'`. The RTDN payload looks like this:

```javascript
{
message: {
data: "eyJ2ZXJzaW9uIjoiMS4wIiwicGFja2FnZU5hbWUiOiJjb20udHdpZ2Fuby5mYXNoaW9uIiwiZXZlbnRUaW1lTWlsbGlzIjoiMTYwNzcwOTkzNTM4MyIsInN1YnNjcmlwdGlvbk5vdGlmaWNhdGlvbiI6eyJ2ZXJzaW9uIjoiMS4wIiwibm90aWZpY2F0aW9uVHlwZSI6NCwicHVyY2hhc2VUb2tlbiI6Im1lZG1tY2dnbmZjYW1ubGlwa2ZlbGluZS5BTy1KMU96aG1Vbk9zNlRKckhCZ3o0MnpNR3BfRklQOXVfS1hRXzZpdjlHTkV6Y3NOcGl3V1U5bzdmYkE3dVdXZnFuLV9lbVdZUFJabkhYX3NjUSIsInN1YnNjcmlwdGlvbklkIjoid2Vla19wcmVtaXVtIn19",
messageId: "1832076433076688",
message_id: "1832076433076688",
publishTime: "2020-12-11T18:05:35.784Z",
publish_time: "2020-12-11T18:05:35.784Z"
},
subscription: "projects/project-id/subscriptions/dev"
}
```

The key `data` is parsed to extract the notification information.

**N.B.** you can not use the provided example, instead, send a test notification from your console, or use a test card to simulate the subscription purchase on your mobile application, use the submitted payload from google for testing.

**Comment by HilalLko** [comment]
Don't know why but I can still see the similar error in logs due to which around 96% of the request to my server throws 500 error.
Even tried to log the exact payload by logging from the file "ServerNotificationController.php"

**
public function google(GoogleDeveloperNotificationRequest $request)
    {
     	$data = $request->getData();
        Log::channel('googlertdn')->info('Inside Google Pay ==>'.$data);
        $developerNotification = DeveloperNotification::parse($data);
        $googleNotification = new GoogleServerNotification($developerNotification);

        if ($googleNotification->isTest()) {
            $version = $developerNotification->getTestNotification()->getVersion();
            Log::info(sprintf("Google Play Test Notification, version: %s", $version));
        }

	if ($developerNotification->isSubscriptionNotification()) {
            $event = GooglePlayEventFactory::create($googleNotification);
            event($event);
        }
    }

**

It logs the same  : 
[2020-12-14 08:51:13] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:19] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:20] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:23] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:26] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:31] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:34] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:38] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:40] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:44] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:50] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:57] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk
[2020-12-14 08:51:58] staging.INFO: Inside Google Pay ==>ZnNkZnNkZnNk

**Reply by imdhemy** [reply]
Is `ZnNkZnNkZnNk` a test payload generated by you, or it's the one sent by google RTDN? This is not a valid payload. kindly, describe the steps to reproduce the error.

**Reply by HilalLko** [reply]
Can you have a look at this answer on StackOverflow

[https://stackoverflow.com/a/50553324](https://stackoverflow.com/questions/50535969/developer-payload-and-server-back-end-interaction-google-in-app-billing/50553324#50553324)

**Reply by HilalLko** [reply]
Not by me, Google is responsible for this

**Reply by HilalLko** [reply]
And for test Notification using MOnetization Setup area is below : 
**PS I have decoded the base 64 encoded string below** 


{"version":"1.0","packageName":"com.developer.hunch","eventTimeMillis":"1607943739507","testNotification":{"version":"1.0"}}

**Comment by HilalLko** [comment]
The problem with the weird Google Pay Payload is simply its pub-sub behavior which sends an empty payload when there are no valid subscriptions available to handle for a particular span of time.
Which is now successfully handled in version [0.8.4](https://github.com/imdhemy/laravel-in-app-purchases/releases/tag/0.8.4) 
responsible code snippet for the fix :
```php
if (! $this->isParsable($data)) {
   Log::info(sprintf("Google Play malformed RTDN: %s", json_encode($request->all())));
   return;
}
```

**Comment by divyame1** [comment]
Hi @imdhemy 
Subscription purchased 
It was running fine some time ago. Suddenly it stopped working. But test notification is working fine, data is also printed in the log file. 
but this listener is not printed the data in a log file. Code lines are...
Please look into this.

```php
<?php

namespace App\Listeners;

use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Queue\InteractsWithQueue;
//use Imdhemy\Purchases\Events\GooglePlay\SubscriptionRenewed;
use Imdhemy\GooglePlay\Subscriptions\SubscriptionPurchase;
use Imdhemy\Purchases\Facades\Subscription;
use Illuminate\Support\Facades\Log;
use Imdhemy\Purchases\Events\PurchaseEvent;
use Imdhemy\Purchases\Contracts\ServerNotificationContract;
use Imdhemy\Purchases\ValueObjects\Time;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionPurchased;
use App\Models\Table;

class SubscribeNotification
{
    protected $event;
    /**
     * Create the event listener.
     *
     * @return void
     */
    public function __construct()
    {
        
    }

    /**
     * Handle the event.
     *
     * @param  Subscribe  $event
     * @return void
     */
    public function handle(SubscriptionPurchased $event)
    {
        Log::info($event);
        $notification = $event->getServerNotification();
        $subscription = $notification->getSubscription();
        $uniqueIdentifier = $subscription->getUniqueIdentifier();
        $skuId = $subscription->getItemId(); 
        $provider = $subscription->getProvider();
        $subscriptionReceipt = Subscription::googlePlay()->id($skuId)->token($uniqueIdentifier)->get();
        // Start Time
        $StartTime = $subscriptionReceipt->getStartTime();
        $data1= $StartTime->toDateTime();
        $dateTime1 = json_encode($data1);
        $Timedate1 =   json_decode($dateTime1);
        $str_time = $Timedate1->date;
        //Expire Time
        $ExpireTime = $subscriptionReceipt->getExpiryTime();
        $data2= $ExpireTime->toDateTime();
        $dateTime2 = json_encode($data2);
        $Timedate2 =   json_decode($dateTime2);
        $exp_time = $Timedate2->date;
        // save in sb
        $data = new Table();
        $data['purchase_token'] = $uniqueIdentifier;
        $data['order_id'] = $subscriptionReceipt->getOrderId();
        $data['start_time'] = $str_time;
        $data['expire_time'] = $exp_time;
        $data['sku_id'] = $skuId;
        $data['price'] = '';
        $data['provider'] = $provider;
        $data['plan_status'] = 'Active';
        $data['renew'] = 'No';
        $data->save();
        Log::info(sprintf("Google Play Purchases, version: %s", 1));
        Log::info(sprintf("Google Play Test Notification, version: %s", $skuId));
        Log::info(sprintf("Google Play Test Notification, version: %s", $uniqueIdentifier));
        Log::info(sprintf("Google Play Test Notification, version: %s", $provider));
        //Log::info(sprintf("Google Play Test Notification, version: %s", $expirationTime));
    }
}
```
Regards
Divya

**Reply by imdhemy** [reply]
Hi @divyame1!

It's a known issue, but it's related to Google Play notifications, not the package. Check The following comment, It may help.

https://github.com/imdhemy/laravel-in-app-purchases/discussions/24#discussioncomment-263768

**Comment by alcaxdeveloper** [comment]
Hi @imdhemy 

I am also using this package and followed all the installation steps mentioned in https://imdhemy.com/laravel-iap-docs/  
But still i am unable to listen  app store notifications .  I don,t  know why. I have tried using both signed and unsigned routes.
Can you please  help me in finding the issue. Currently i am using the sandbox mode.

My code is as follows:-

<?php

use Imdhemy\Purchases\Events\AppStore\Cancel;
use Imdhemy\Purchases\Events\AppStore\DidChangeRenewalPref;
use Imdhemy\Purchases\Events\AppStore\DidChangeRenewalStatus;
use Imdhemy\Purchases\Events\AppStore\DidFailToRenew;
use Imdhemy\Purchases\Events\AppStore\DidRecover;
use Imdhemy\Purchases\Events\AppStore\DidRenew;
use Imdhemy\Purchases\Events\AppStore\InitialBuy;
use Imdhemy\Purchases\Events\AppStore\InteractiveRenewal;
use Imdhemy\Purchases\Events\AppStore\PriceIncreaseConsent;
use Imdhemy\Purchases\Events\AppStore\Refund;
use Imdhemy\Purchases\Events\AppStore\Revoke;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionCanceled;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionDeferred;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionExpired;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionInGracePeriod;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionOnHold;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionPaused;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionPauseScheduleChanged;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionPriceChangeConfirmed;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionPurchased;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionRecovered;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionRenewed;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionRestarted;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionRevoked;

return [
  'routing' => [
    'signed' => false,
  ],

  'google_play_package_name' => env('GOOGLE_PLAY_PACKAGE_NAME', 'com.some.thing'),

  'appstore_password' => env('APPSTORE_PASSWORD', 'XXXXXXX'),
  'appstore_sandbox' => env('APPSTORE_SANDBOX', true),

  'eventListeners' => [
      /**
       * --------------------------------------------------------
       * Google Play Events
       * --------------------------------------------------------
       */
    SubscriptionPurchased::class => [],
    SubscriptionRenewed::class => [],
    SubscriptionInGracePeriod::class => [],
    SubscriptionExpired::class => [],
    SubscriptionCanceled::class => [],
    SubscriptionPaused::class => [],
    SubscriptionRestarted::class => [],
    SubscriptionDeferred::class => [],
    SubscriptionRevoked::class => [],
    SubscriptionOnHold::class => [],
    SubscriptionRecovered::class => [],
    SubscriptionPauseScheduleChanged::class => [],
    SubscriptionPriceChangeConfirmed::class => [],

      /**
       * --------------------------------------------------------
       * App Store Events
       * --------------------------------------------------------
       */
    Cancel::class => [\App\Listeners\Subscription\Apple\SubscriptionCanceledListener::class],
    DidChangeRenewalPref::class => [],
    DidChangeRenewalStatus::class => [],
    DidFailToRenew::class => [],
    DidRecover::class => [\App\Listeners\Subscription\Apple\SubscriptionDidRecoverListener::class],
    DidRenew::class => [\App\Listeners\Subscription\Apple\SubscriptionDidRenewListener::class],
    InitialBuy::class => [\App\Listeners\Subscription\Apple\SubscriptionPurchasedListener::class],
    InteractiveRenewal::class => [],
    PriceIncreaseConsent::class => [],
    Refund::class => [],
    Revoke::class => [],
  ],
];

Code of one of my listener is as follows:-

<?php

namespace App\Listeners\Subscription\Apple;

use Illuminate\Support\Facades\Log;
use Imdhemy\Purchases\Events\AppStore\DidRenew;
// use App\Models\UserSubscriptionHistory;

class SubscriptionDidRenewListener
{
    /**
     * Handle the event.
     *
     * @param DidRenew $event
     * @return void
     */

    public function handle(DidRenew $event)
    {
        //  The following data can be retrieved from the event
        try {
            $notification = $event->getServerNotification()->getSubscription();
            
            $uniqueIdentifier = $subscription->getUniqueIdentifier();
            $expirationTime = $subscription->getExpiryTime();

            $item_id = $subscription->getItemId();
             Log::info(array("data"=> $notification,"uniqueIdentifier"=> $uniqueIdentifier,"expirationTime"=> $expirationTime,"item_id"=> $item_id));
             $subscription = $notification->getSubscription();
              Log::info(array("subscription"=> $subscription,'name'=>"didrenew"));
            // $data = UserSubscriptionHistory::googleHisUpdateOrCreate($subscription, $item_id, 'canceled');
            // if ($data) {
            //     UserSubscriptionHistory::SubscriptionFirebaseNotification($data, 'canceled');
            // }
        } catch(\Exception $e){
            Log::info($e->getMessage());
        }

    }
}

But no notification is received. Can you please help me to debug the issue.

Thanks

**Reply by HilalLko** [reply]
@alcaxdeveloper 
Kindly reformat your question using code block inside '```' so that it can be easily readable and viewable :-)

Example :
```
Your codes here...
```

**Reply by alcaxdeveloper** [reply]
Hi @imdhemy Thanks for your quick reply. Kindly check  Formatted code:-

I am also using this package and followed all the installation steps mentioned in https://imdhemy.com/laravel-iap-docs/
But still i am unable to listen app store notifications . I don,t know why. I have tried using both signed and unsigned routes.
Can you please help me in finding the issue. Currently i am using the sandbox mode.

My code is as follows of config>liap.php:-

`<?php

use Imdhemy\Purchases\Events\AppStore\Cancel;
use Imdhemy\Purchases\Events\AppStore\DidChangeRenewalPref;
use Imdhemy\Purchases\Events\AppStore\DidChangeRenewalStatus;
use Imdhemy\Purchases\Events\AppStore\DidFailToRenew;
use Imdhemy\Purchases\Events\AppStore\DidRecover;
use Imdhemy\Purchases\Events\AppStore\DidRenew;
use Imdhemy\Purchases\Events\AppStore\InitialBuy;
use Imdhemy\Purchases\Events\AppStore\InteractiveRenewal;
use Imdhemy\Purchases\Events\AppStore\PriceIncreaseConsent;
use Imdhemy\Purchases\Events\AppStore\Refund;
use Imdhemy\Purchases\Events\AppStore\Revoke;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionCanceled;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionDeferred;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionExpired;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionInGracePeriod;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionOnHold;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionPaused;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionPauseScheduleChanged;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionPriceChangeConfirmed;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionPurchased;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionRecovered;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionRenewed;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionRestarted;
use Imdhemy\Purchases\Events\GooglePlay\SubscriptionRevoked;

return [
  'routing' => [
    'signed' => false,
  ],

  'google_play_package_name' => env('GOOGLE_PLAY_PACKAGE_NAME', 'com.some.thing'),

  'appstore_password' => env('APPSTORE_PASSWORD', 'XXXXXXX'),
  'appstore_sandbox' => env('APPSTORE_SANDBOX', true),

  'eventListeners' => [
      /**
       * --------------------------------------------------------
       * Google Play Events
       * --------------------------------------------------------
       */
    SubscriptionPurchased::class => [],
    SubscriptionRenewed::class => [],
    SubscriptionInGracePeriod::class => [],
    SubscriptionExpired::class => [],
    SubscriptionCanceled::class => [],
    SubscriptionPaused::class => [],
    SubscriptionRestarted::class => [],
    SubscriptionDeferred::class => [],
    SubscriptionRevoked::class => [],
    SubscriptionOnHold::class => [],
    SubscriptionRecovered::class => [],
    SubscriptionPauseScheduleChanged::class => [],
    SubscriptionPriceChangeConfirmed::class => [],

      /**
       * --------------------------------------------------------
       * App Store Events
       * --------------------------------------------------------
       */
    Cancel::class => [\App\Listeners\Subscription\Apple\SubscriptionCanceledListener::class],
    DidChangeRenewalPref::class => [],
    DidChangeRenewalStatus::class => [],
    DidFailToRenew::class => [],
    DidRecover::class => [\App\Listeners\Subscription\Apple\SubscriptionDidRecoverListener::class],
    DidRenew::class => [\App\Listeners\Subscription\Apple\SubscriptionDidRenewListener::class],
    InitialBuy::class => [\App\Listeners\Subscription\Apple\SubscriptionPurchasedListener::class],
    InteractiveRenewal::class => [],
    PriceIncreaseConsent::class => [],
    Refund::class => [],
    Revoke::class => [],
  ],
];
`
and code for one of my listeners is as follows

`<?php

namespace App\Listeners\Subscription\Apple;

use Illuminate\Support\Facades\Log;
use Imdhemy\Purchases\Events\AppStore\DidRenew;
// use App\Models\UserSubscriptionHistory;

class SubscriptionDidRenewListener
{
    /**
     * Handle the event.
     *
     * @param DidRenew $event
     * @return void
     */

    public function handle(DidRenew $event)
    {
        //  The following data can be retrieved from the event
        try {
            $notification = $event->getServerNotification()->getSubscription();
            
            $uniqueIdentifier = $subscription->getUniqueIdentifier();
            $expirationTime = $subscription->getExpiryTime();

            $item_id = $subscription->getItemId();
             Log::info(array("data"=> $notification,"uniqueIdentifier"=> $uniqueIdentifier,"expirationTime"=> $expirationTime,"item_id"=> $item_id));
             $subscription = $notification->getSubscription();
              Log::info(array("subscription"=> $subscription,'name'=>"didrenew"));
            // $data = UserSubscriptionHistory::googleHisUpdateOrCreate($subscription, $item_id, 'canceled');
            // if ($data) {
            //     UserSubscriptionHistory::SubscriptionFirebaseNotification($data, 'canceled');
            // }
        } catch(\Exception $e){
            Log::info($e->getMessage());
        }

    }
}
`
But no notification is received. Can you please help me to debug the issue. Thanks

Formatted code.

**Reply by HilalLko** [reply]
@alcaxdeveloper 
try to add Logger just before the beginning of try block kind'f
```
Log::info('Inside AppStore Auto Renew Subscription Listener ===> '.print_r($event,true));
```

**Reply by alcaxdeveloper** [reply]
> @alcaxdeveloper try to add Logger just before the beginning of try block kind'f
> 
> ```
> Log::info('Inside AppStore Auto Renew Subscription Listener ===> '.print_r($event,true));
> ```
@imdhemy 

Thanks for your reply !
Tried this approach but still there are no logs.

**Reply by HilalLko** [reply]
@alcaxdeveloper 
This means that your RTDN notifications are not handled by the Laravel app yet, try to recheck the configuration.

[check here](https://imdhemy.com/laravel-iap-docs/docs/server-notifications/app-store)

**Reply by alcaxdeveloper** [reply]
@hilal-hipster 

I have checked my configuration again every thing is configured properly.

My configured url in app store  https://healthelivin.alcax.com/liap/notifications?provider=app-store.

Please help.

Thanks

**Comment by imdhemy** [comment]
@alcaxdeveloper 

Make sure you did the following:

- [ ] Using LIAP v`1.0` or higher
- [ ] Check if the notification endpoint is enabled `php artisan route:list`
- [ ] You are using an `https` protocol for the generated URL. `https://healthelivin.alcax.com/liap/notifications?provider=app-store`
- [ ] You followed the instructions in [documentation](https://imdhemy.com/laravel-iap-docs/docs/server-notifications/app-store).
- [ ] You deployed to a live server 😅
- [ ] You are using Notifications v.1 from App Store
- [ ] You configured the [sandbox tester accounts](https://developer.apple.com/documentation/storekit/in-app_purchase/testing_in-app_purchases_with_sandbox).

---
Side Notes:

- REMOVE your App Store password from your question/comments here, this is not safe.
- Don't wait for the initial purchase event, most probably it will not be sent.

**Reply by alcaxdeveloper** [reply]
@imdhemy 

For point !:-  in my terminal while getting list of installed package it is mentioned that  imdhemy/appstore-iap               1.0.0     PHP Appstore In-App Purchase implementation  >>> okay
For point 2:-   in my route list there is a route  as POST      liap/notifications liap.serverNotifications › Imdhemy\Purchases   >>> okay
For point 3:- Tested the url in postman the url is working.
For point 4:-  all the instructions are followed only (If your server requires IP addresses to be on an allow list, add the IP address subnet 17.0.0.0/8 to allow your server to receive notifications from the App Store server.) this one is not clear to me. Is this mandatory.
For point 6:- yes i am using v1 notification in app store. >> okay
For point 7:-  Yes we have configured the tester account.

And also i am checking for did renew event in sandbox also.
But no notification is received on server.

Please help.
Thanks

**Reply by imdhemy** [reply]
@alcaxdeveloper Which type of products you are purchasing on your test?

**Reply by alcaxdeveloper** [reply]
> Which type of products you are purchasing on your test?

the type is auto-renewable subsciption.

**Reply by imdhemy** [reply]
@alcaxdeveloper 

Can you change the URL submitted to the App Store to another URL you can log/debug? 
To be sure what the root of the problem is, is it related to the LIAP package, or does Apple not send you any notification?

It would be better if you are testing a weekly subscription.

**Reply by alcaxdeveloper** [reply]
@imdhemy 

Thanks for your quick support and for your best efforts. I did the testing in same way and found that notifications are working directly from apple server.Which means apple is sending notifications directly. The issue is related with LIAP Package.

Thanks

**Reply by imdhemy** [reply]
@alcaxdeveloper 

Sounds good, now we have something to dig in.

Could you show me the request payload you receive from Apple on that endpoint?

You can do something like:

```php
dd(request()->all());
```

Or any other suitable idea.

**Reply by imdhemy** [reply]
@alcaxdeveloper 

Or at least confirm which key you see, 

- A single key `signedPayload`
- Multiple keys along with `unified_receipt`

**Reply by amrish2020** [reply]
Hi @imdhemy 

I am using this package, to get notifications but it's not working it's not firing or calling any event,

- Using LIAP v1.0 or higher
  -- Using LIAP 1.30, Laravel 8.75
- Check if the notification endpoint is enabled php artisan route:list
  -- yes i have checked the route list and it showing fine
  POST     | liap/notifications                        | liap.serverNotifications  | Imdhemy\Purchases\Http\Controllers\ServerNotificationController           | web  
- You are using an https protocol for the generated URL.
 -- yes i am using it with signed route https://<domain>/liap/notifications?signature=<signature>&provider=app-store
 - You followed the instructions in.
  -- I have followed and already added a test user and created a Listener class
- You deployed to a live server 😅
  -- yes its in the live server
 - You are using Notifications v.1 from App Store
  -- In current version, there is no option to select it,
 - You configured the [sandbox tester accounts](https://developer.apple.com/documentation/storekit/in-app_purchase/testing_in-app_purchases_with_sandbox).
  -- yes I have created the test account.
  
Kindly please help to resolve this, will be much appreciated.

Thank you
